### PR TITLE
driver: sensor: Add support for LIS2MDL Mag sensor

### PR DESCRIPTION
**Pull request by mariotesi**
Added support to LIS2MDL Magnetometer sensor provided
with following features:
	- I2C interface
	- Mag data
	- Temperature data
	- ODR configurable by config or at runtime
	- Trigger mode selectable by menuconfig
	- IRQ pin configurable
	- Hard Iron offset setting at runtime

Tested on ST MEMS IKS01A2 + NUCLEO STM32F411RE board.
LIS2MDL connected to I2C master interface (SPI 3 wire not
supported yet).
Test run with all ODR {10, 20, 50, 100} Hz in poll and
trigger mode.

For more info on this LIS2MDL please follow this link:
http://www.st.com/en/mems-and-sensors/lis2mdl.html

Signed-off-by: Mario Tesi <mario.tesi@st.com>

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Need device tree binding/yaml for these sensor.  Use compatiable “st,lis2mdl”.  See dts/bindings/sensor/st,lis3mdl-magn.yaml for example.

File: drivers/sensor/lis2mdl/Kconfig [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+menuconfig LIS2MDL
+	bool
+	prompt "LIS2MDL Magnetometer"
+	depends on I2C
+	default n
+	help
+	  Enable driver for LIS2MDL I2C-based magnetometer sensor.
+
+config LIS2MDL_NAME
```
Comment (erwango) [inline drivers/sensor/lis2mdl/Kconfig]: CONFIG_LIS2MDL_I2C_MASTER_DEV_NAME, CONFIG_LIS2MDL_I2C_ADDR and LIS2MDL_I2C_MASTER_DEV_NAME will be generated from device tree "compilation".
Hence in order to prevent having them genetared twice, you should protect them under ```if !HAS_DTS_I2C_DEVICE``` directive.
You can check other drivers like LIS3MDL to check how this should look like.

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
One remaining comment, otherwise looks fine.

**Review by erwango — APPROVED** [review]

**Comment by mariotesi** [comment]
There are still some issue on this commit ?
Please let me know

**Comment by erwango** [comment]
@galak, can you review again? I think this is ok now (binding file added)

**Comment by avisconti** [comment]
@galak 
Can you possibly take a look at this PR?

**Review by galak — DISMISSED** [review]
Can we change the GPIO for interrupt to use DT

**Comment by mariotesi** [comment]
In the last commit device tree support has been added to GPIO interrupt selection pin/port

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+	u8_t regs = LIS2MDL_OFFSET_X_REG_L;
+	u8_t i;
+	s16_t offs;
+	int ret;
+
+	for (i = 0; i < LIS2MDL_NUM_AXIS; i++) {
+		offs = val->val1;
+		ret = lis2mdl_raw_write_n(lis2mdl, CONFIG_LIS2MDL_I2C_ADDR,
+					  regs, (u8_t *)&offs,
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: You assume the sensor and MCU have the same endianness. Use a byte order macro in `include/misc/byteorder.h`

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+				     struct sensor_value *val)
+{
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+	u16_t temp_raw;
+	s32_t temp;
+	int ret;
+
+	ret = lis2mdl_raw_read_n(lis2mdl, CONFIG_LIS2MDL_I2C_ADDR,
+				 LIS2MDL_TEMP_OUT_L_REG,
+				 (u8_t *)&temp_raw, sizeof(temp_raw));
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: You should be doing the i2c transfer in a sample_fetch routine, not channel_get.

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+				struct sensor_value *val)
+{
+	switch (chan) {
+	case SENSOR_CHAN_MAGN_X:
+	case SENSOR_CHAN_MAGN_Y:
+	case SENSOR_CHAN_MAGN_Z:
+	case SENSOR_CHAN_MAGN_XYZ:
+		lis2mdl_channel_get_mag(dev, chan, val);
+		return 0;
+	case SENSOR_CHAN_TEMP:
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: This should be `SENSOR_CHAN_DIE_TEMP`

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+			s16_t m_axis[3];
+		};
+	} buf __aligned(2);
+
+	if (lis2mdl_raw_read8(drv_data, CONFIG_LIS2MDL_I2C_ADDR,
+			      LIS2MDL_STATUS_REG, &tmp)) {
+		return -EIO;
+	}
+
+	if (!(tmp & LIS2MDL_STS_MDA_UP)) {
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: What is this checking?

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+#endif
+	.sample_fetch = lis2mdl_sample_fetch,
+	.channel_get = lis2mdl_channel_get,
+};
+
+static int lis2mdl_init_interface(struct device *dev)
+{
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+
+	lis2mdl->i2c = device_get_binding(CONFIG_LIS2MDL_I2C_MASTER_DEV_NAME);
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: Store the i2c dev name in a const device config struct and get it from there.

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+	lis2mdl->mag_fs_sensitivity = LIS2MDL_SENSITIVITY;
+
+#ifdef CONFIG_LIS2MDL_TRIGGER
+	if (lis2mdl_init_interrupt(dev) < 0) {
+		SYS_LOG_DBG("Failed to initialize interrupts");
+		return -EIO;
+	}
+#endif
+
+	dev->driver_api = &lis2mdl_driver_api;
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: Remove this line and use `DEVICE_AND_API_INIT` below.

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+	struct sensor_trigger drdy_trigger = {
+		.type = SENSOR_TRIG_DATA_READY,
+	};
+
+	if (lis2mdl->handler_drdy != NULL) {
+		lis2mdl->handler_drdy(dev, &drdy_trigger);
+	}
+
+	gpio_pin_enable_callback(lis2mdl->gpio, CONFIG_LIS2MDL_GPIO_PIN_NUM);
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl_trigger.c]: Get the pin num from the const device config struct

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	lis2mdl_handle_interrupt(lis2mdl->dev);
+}
+#endif
+
+int lis2mdl_init_interrupt(struct device *dev)
+{
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+
+	/* setup data ready gpio interrupt */
+	lis2mdl->gpio = device_get_binding(CONFIG_LIS2MDL_GPIO_DEV_NAME);
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl_trigger.c]: Get the gpio device name from the const device config struct

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
@mariotesi , do you have further plans for this PR?

**Comment by mariotesi** [comment]
I'll push CR asap

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add the new sensor driver to `tests/drivers/build_all\sensors_i_z.conf` and `tests/drivers/build_all\sensors_trigger_i_z.conf` to ensure it gets built in CI.

File: drivers/sensor/lis2mdl/lis2mdl.h [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+int lis2mdl_trigger_set(struct device *dev,
+			  const struct sensor_trigger *trig,
+			  sensor_trigger_handler_t handler);
+#endif /* CONFIG_LIS2MDL_TRIGGER */
+
+int lis2mdl_write_data_with_mask(struct lis2mdl_data *lis2mdl, u8_t reg_addr,
+				   u8_t mask, u8_t data);
+int lis2mdl_raw_read8(struct lis2mdl_data *lis2mdl, u8_t reg_addr, u8_t *value);
+int lis2mdl_raw_write8(struct lis2mdl_data *lis2mdl, u8_t reg_addr, u8_t value);
+int lis2mdl_raw_read_n(struct lis2mdl_data *lis2mdl, u8_t start_addr, u8_t *buf,
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.h]: These i2c wrapper prototypes can now be removed

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+static int lis2mdl_set_odr(struct device *dev, u16_t odr)
+{
+	int odr_idx;
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+
+	/* check if power off */
+	if (odr == 0) {
+		/* power off mag */
+		return i2c_reg_update_byte(lis2mdl->i2c,
+					   CONFIG_LIS2MDL_I2C_ADDR,
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: Get the i2c address from the device struct

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+					   LIS2MDL_MAG_MODE_MASK,
+					   LIS2MDL_MD_IDLE1_MODE);
+	}
+
+	odr_idx = lis2mdl_freq_to_odr_val(odr);
+	if (odr_idx < 0) {
+		return odr_idx;
+	}
+
+	return i2c_reg_update_byte(lis2mdl->i2c, CONFIG_LIS2MDL_I2C_ADDR,
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: Get the i2c address from the device struct

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please fix the CI failures. It looks like you need to add defines to `tests/drivers/build_all/dts.fixup`

**Comment by mariotesi** [comment]
I don't understand why shippable test fails, my build of test/drivers/build_all everything it's ok

**Comment by MaureenHelm** [comment]
> I don't understand why shippable test fails, my build of test/drivers/build_all everything it's ok

How are you building it locally? Did you include `-DCONF_FILE=sensors_i_z.conf` when you invoked cmake? There are two ways you should build the test:
```console
$ cmake -DBOARD=frdm_k64f -DCONF_FILE=sensors_i_z.conf ..
$ make
```
```console
$ cmake -DBOARD=frdm_k64f -DCONF_FILE=sensors_trigger_i_z.conf ..
$ make
```

File: tests/drivers/build_all/dts.fixup [hunk 13]
```diff
@@ -105,6 +105,16 @@
 #define CONFIG_VL53L0X_I2C_MASTER_DEV_NAME ""
 #endif
 
+#ifndef CONFIG_LIS2MDL_I2C_ADDR
+#define CONFIG_LIS2MDL_I2C_ADDR 0
+#define CONFIG_LIS2MDL_I2C_MASTER_DEV_NAME ""
+#endif
```
Comment (MaureenHelm) [inline tests/drivers/build_all/dts.fixup]: Please move the tests changes into a separate commit.

File: tests/drivers/build_all/sensors_i_z.conf [hunk 14]
```diff
@@ -28,3 +28,4 @@ CONFIG_TH02=y
 CONFIG_TMP007=y
 CONFIG_TMP112=y
 CONFIG_VL53L0X=y
+CONFIG_LIS2MDL=y
```
Comment (MaureenHelm) [inline tests/drivers/build_all/sensors_i_z.conf]: Please move this up to preserve alpha order

File: tests/drivers/build_all/sensors_trigger_i_z.conf [hunk 15]
```diff
@@ -25,3 +25,5 @@ CONFIG_SX9500=y
 CONFIG_SX9500_TRIGGER_OWN_THREAD=y
 CONFIG_TMP007=y
 CONFIG_TMP007_TRIGGER_OWN_THREAD=y
+CONFIG_LIS2MDL=y
+CONFIG_LIS2MDL_TRIGGER_OWN_THREAD=y
```
Comment (MaureenHelm) [inline tests/drivers/build_all/sensors_trigger_i_z.conf]: Please move up to preserve alpha order

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by MaureenHelm — APPROVED** [review]
